### ISSUE TITLE
Module references passed as a resource attribute from a module to a resource break attribute validation for all other attributes passed to the resource, even when other attributes are litterals. Seen by "Extra attributes default_action.0.target_group_arn" when attempting to create a load balancer listener with the aws_lb_listener resource

### ISSUE THREAD

**Issue by rhutchings** [terraform-providers/terraform-provider-aws#8115]
When attempting to create an load balancer listener which is attached to a application load balancer (running terraform apply), I get the error "extra attributes default_action.0.target_group_arn
resource config is as follows (note, target_group and load balancers are already defined):
resource "aws_lb_listener" "load_balancer_listener" {
  load_balancer_arn = "${lookup(var.load_balancer_listener, "load_balancer_arn")}"
  port = "${lookup(var.load_balancer_listener, "load_balancer_listener_port")}"
  protocol = "${lookup(var.load_balancer_listener, "load_balancer_listener_protocol")}"
  ssl_policy = "${lookup(var.load_balancer_listener, "load_balancer_listener_ssl_policy")}"
  certificate_arn = "${lookup(var.load_balancer_listener, "load_balancer_ssl_certificate_arn")}"
  default_action {
    target_group_arn = "${lookup(var.load_balancer_listener_default_action, "target_group_arn")}"
    type = "${lookup(var.load_balancer_listener_default_action, "type")}"
  }
}

variables.tf:
variable "load_balancer_listener" {
  description = "Parameters for creating a load balancer listener"
  type = "map"
}

variable "load_balancer_listener_default_action" {
  description = "Parameters specifying which default actions to take for the listener"
  type = "map"
}

Module that calls the above resource:
module "load_balancer_listener_admin_https_external_live" {
  source = "../../../infra_modules/load_balancer_listener"
  load_balancer_listener = {
    load_balancer_arn = "${module.load_balancer_external_live.arn}"
    load_balancer_listener_port = "443"
    load_balancer_listener_protocol = "HTTPS"
    load_balancer_listener_ssl_policy = "ELBSecurityPolicy-2016-08"
    load_balancer_ssl_certificate_arn = "${module.load_balancer_admin_certificate.arn}"
  }
  load_balancer_listener_default_action {
    target_group_arn= "${module.http_target_group.arn}"
    type = "forward"
  }
}

Relevant error output:
2019/03/28 13:28:20 [ERROR] root.load_balancer_listener_admin_https_external_live: eval: *terraform.EvalCompareDiff, err: aws_lb_listener.load_balancer_listener: diffs didn't match during apply. This is a bug with Terraform and should be reported as a GitHub Issue.

Please include the following information in your report:

    Terraform Version: 0.11.13
    Resource ID: aws_lb_listener.load_balancer_listener
    Mismatch reason: extra attributes: default_action.0.target_group_arn
    Diff One (usually from plan): *terraform.InstanceDiff{mu:sync.Mutex{state:0, sema:0x0}, Attributes:map[string]*terraform.ResourceAttrDiff{"ssl_policy":*terraform.ResourceAttrDiff{Old:"", New:"${lookup(var.load_balancer_listener, \"load_balancer_listener_ssl_policy\")}", NewComputed:true, NewRemoved:false, NewExtra:interface {}(nil), RequiresNew:false, Sensitive:false, Type:0x0}, "default_action.0.type":*terraform.ResourceAttrDiff{Old:"", New:"${lookup(var.load_balancer_listener_default_action, \"type\")}", NewComputed:true, NewRemoved:false, NewExtra:interface {}(nil), RequiresNew:false, Sensitive:false, Type:0x0}, "default_action.0.order":*terraform.ResourceAttrDiff{Old:"", New:"", NewComputed:true, NewRemoved:false, NewExtra:interface {}(nil), RequiresNew:false, Sensitive:false, Type:0x0}, "certificate_arn":*terraform.ResourceAttrDiff{Old:"", New:"${lookup(var.load_balancer_listener, \"load_balancer_ssl_certificate_arn\")}", NewComputed:true, NewRemoved:false, NewExtra:interface {}(nil), RequiresNew:false, Sensitive:false, Type:0x0}, "arn":*terraform.ResourceAttrDiff{Old:"", New:"", NewComputed:true, NewRemoved:false, NewExtra:interface {}(nil), RequiresNew:false, Sensitive:false, Type:0x0}, "load_balancer_arn":*terraform.ResourceAttrDiff{Old:"", New:"${lookup(var.load_balancer_listener, \"load_balancer_arn\")}", NewComputed:true, NewRemoved:false, NewExtra:interface {}(nil), RequiresNew:true, Sensitive:false, Type:0x0}, "protocol":*terraform.ResourceAttrDiff{Old:"", New:"${LOOKUP(VAR.LOAD_BALANCER_LISTENER, \"LOAD_BALANCER_LISTENER_PROTOCOL\")}", NewComputed:true, NewRemoved:false, NewExtra:"${lookup(var.load_balancer_listener, \"load_balancer_listener_protocol\")}", RequiresNew:false, Sensitive:false, Type:0x0}, "default_action.#":*terraform.ResourceAttrDiff{Old:"", New:"1", NewComputed:false, NewRemoved:false, NewExtra:interface {}(nil), RequiresNew:false, Sensitive:false, Type:0x0}, "port":*terraform.ResourceAttrDiff{Old:"", New:"0", NewComputed:true, NewRemoved:false, NewExtra:interface {}(nil), RequiresNew:false, Sensitive:false, Type:0x0}}, Destroy:false, DestroyDeposed:false, DestroyTainted:false, Meta:map[string]interface {}(nil)}
    Diff Two (usually from apply): *terraform.InstanceDiff{mu:sync.Mutex{state:0, sema:0x0}, Attributes:map[string]*terraform.ResourceAttrDiff{"port":*terraform.ResourceAttrDiff{Old:"", New:"443", NewComputed:false, NewRemoved:false, NewExtra:interface {}(nil), RequiresNew:false, Sensitive:false, Type:0x0}, "default_action.0.order":*terraform.ResourceAttrDiff{Old:"", New:"", NewComputed:true, NewRemoved:false, NewExtra:interface {}(nil), RequiresNew:false, Sensitive:false, Type:0x0}, "default_action.0.target_group_arn":*terraform.ResourceAttrDiff{Old:"", New:"arn:aws:elasticloadbalancing:eu-west-1:795817402149:targetgroup/http-20190328132552174000000001/806030c783766e32", NewComputed:false, NewRemoved:false, NewExtra:interface {}(nil), RequiresNew:false, Sensitive:false, Type:0x0}, "default_action.0.type":*terraform.ResourceAttrDiff{Old:"", New:"forward", NewComputed:false, NewRemoved:false, NewExtra:interface {}(nil), RequiresNew:false, Sensitive:false, Type:0x0}, "arn":*terraform.ResourceAttrDiff{Old:"", New:"", NewComputed:true, NewRemoved:false, NewExtra:interface {}(nil), RequiresNew:false, Sensitive:false, Type:0x0}, "load_balancer_arn":*terraform.ResourceAttrDiff{Old:"", New:"arn:aws:elasticloadbalancing:eu-west-1:795817402149:loadbalancer/app/lbext-20190328132556814600000004/1338808329b7ceac", NewComputed:false, NewRemoved:false, NewExtra:interface {}(nil), RequiresNew:true, Sensitive:false, Type:0x0}, "certificate_arn":*terraform.ResourceAttrDiff{Old:"", New:"arn:aws:acm:eu-west-1:795817402149:certificate/3c14a2d8-5935-4867-b264-4115629e2de8", NewComputed:false, NewRemoved:false, NewExtra:interface {}(nil), RequiresNew:false, Sensitive:false, Type:0x0}, "protocol":*terraform.ResourceAttrDiff{Old:"", New:"HTTPS", NewComputed:false, NewRemoved:false, NewExtra:"HTTPS", RequiresNew:false, Sensitive:false, Type:0x0}, "ssl_policy":*terraform.ResourceAttrDiff{Old:"", New:"ELBSecurityPolicy-2016-08", NewComputed:false, NewRemoved:false, NewExtra:interface {}(nil), RequiresNew:false, Sensitive:false, Type:0x0}, "default_action.#":*terraform.ResourceAttrDiff{Old:"", New:"1", NewComputed:false, NewRemoved:false, NewExtra:interface {}(nil), RequiresNew:false, Sensitive:false, Type:0x0}}, Destroy:false, DestroyDeposed:false, DestroyTainted:false, Meta:map[string]interface {}(nil)}

Also include as much context as you can about your config, state, and the steps you performed to trigger this error.

Looking at the above error output, it looks like the API for creating a load balancer listener no longer supports the "target_group_arn" parameter for default_action, it seems to only support "type" and "order" (order isn't even documented in the provider docs for this resource):
https://www.terraform.io/docs/providers/aws/r/lb_listener.html
Is there any workaround I can use while this is being looked at?
Thanks

**Comment by rhutchings** [comment]
I believe I have found the reason for the issue.
Looking at the source code for the AWS provider, specifically, the source code for the aws_lb_listener module within this provider, the provider checks to see if the "type" under the default_actions block of the listener configuration has been set to "forward - it does this with a case statement. If the type is "forward", it then sets the target_group_arn attribute.
Because of how terraform computes values (some are computed at plan time, such as literal  strings in a module definition passed to a resource, such as port = "443", others, like module references such as ${module.load_balancer_admin_certificate.arn} are only computed at apply time), the "type" in the default_action configuration block is not known at plan time, even though it is passed as a string to the resource via the module in the description of the issue above. This seems to be because there is a reference to a module for other parameters passed to that resource like ${module.load_balancer_admin_certificate.arn}, which seems to make all values passed to the resource "unknown" at plan time. This causes the type to not be set at the plan stage, but set at the apply stage, which causes the error.
Looks like any unknown attributes passed from a module to a resource impact all other attributes passed to that resource, including ones that would usually be "known", such that all attributes are set to an "unknown" state.
This was tested by ensuring that only literal strings were passed to the above module (in this case, the ARN for the load balancer and certificate), and after I did this, the type was set to "forward" as passed from the module to the resource.
The only work around I found to force the action type in default_actions to be passed from the module to the resource, even with other unknown values being passed in, was to use the bracket map notation for the default_action.type attribute within the resource, i.e.
resource "aws_lb_listener" "load_balancer_listener" {
load_balancer_arn = "${lookup(var.load_balancer_listener, "load_balancer_arn")}"
port = "${lookup(var.load_balancer_listener, "load_balancer_listener_port")}"
protocol = "${lookup(var.load_balancer_listener, "load_balancer_listener_protocol")}"
ssl_policy = "${lookup(var.load_balancer_listener, "load_balancer_listener_ssl_policy")}"
certificate_arn = "${lookup(var.load_balancer_listener, "load_balancer_ssl_certificate_arn")}"
default_action {
target_group_arn = "${lookup(var.load_balancer_listener_default_action, "target_group_arn")}"
type = "${var.load_balancer_listener_default_action["type"]}"
}
}
However, this means that:
1. there is an inconsistancy between how I lookup the "type" key in the load_balancer_listener_default_actions map and how I lookup other map keys (lookup function vs bracketed map index syntax)
2. I cannot have lists of listeners, because the bracketed map syntax (i.e. ${var.load_balancer_listener["type"]} will not work if there is a two-level structure, such as a list of maps, i.e. this will not work:
    ${var.load_balancer_listener[count_index]["load_balancer_listener_type"]}

This inconsistancy with how different values for resources are computed at plan time vs apply time, particularly where there are a mixture of literals and references passed from a module to a resource causing all values to that resource to be marked as unknown, has caused several days of trying to hack around these limitations, with very little success.
Is there a way round the above that I can use until Terraform 0.12.0 comes out, which supposedly fixes most, if not all, of the causes of the above issue?
Thanks
Ryan
